### ISSUE TITLE
Implement dropout layer

### ISSUE THREAD

**Issue by milancurcic** [modern-fortran/neural-fortran#170]
References: 

* [Keras](https://keras.io/api/layers/regularization_layers/dropout/)
* [PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html)
* [ATHENA](https://github.com/nedtaylor/athena/blob/main/src/lib/mod_dropout_layer.f90)

**Comment by ricor07** [comment]
Hello, i'd like to work on this

**Comment by milancurcic** [comment]
Hi @ricor07, that'd be fantastic, thank you!

* To start, take a look at how a concrete layer type extends the abstract base layer, for example [`nf_dense_layer`](https://github.com/modern-fortran/neural-fortran/blob/main/src/nf/nf_dense_layer.f90). Then follow the same pattern by defining a `nf_dropout_layer` type in `src/nf/nf_dropout_layer.f90`.
* `nf_dropout_layer` will need the `init`, `forward`, and `backward` methods.
* In [nf_layer_constructors](https://github.com/modern-fortran/neural-fortran/blob/main/src/nf/nf_layer_constructors.f90), you'll need to add a constructor function `dropout` that returns an instance of `nf_dropout_layer` type. You can model this function based on the corresponding counterparts in Keras and PyTorch.
* Create a test program in `test/test_dropout_layer.f90`. We'd first want to ensure that the dropout layer behaves the way we expect (e.g. creating a dropout layer given inputs should have a specific known shape etc.).
* Once we're happy with the above, we can proceed to add the forward and backward invocations of dropout in the `network % forward` and `network % backward` implementations in [nf_network_submodule](https://github.com/modern-fortran/neural-fortran/blob/main/src/nf/nf_network_submodule.f90).

Let me know if you have any questions. If you'd like, we could also meet on video to discuss this after January 6.

**Comment by ricor07** [comment]
Thank you for your kindness. I'll start to work.

**Comment by ricor07** [comment]
@milancurcic Sorry for disturbing you, but i can't compile the repo. I downloaded fpm and gfortran, i included them in the path as well. After running fpm build --profile release, this is what I get:

`PS C:\Users\angel\neural-fortran> fpm build --profile release
 + mkdir build\dependencies
nf_random.f90                          done.                                                                                                                                                                                           
nf_optimizers.f90                      done.
nf_parallel.f90                        done.
nf_base_layer.f90                      done.
nf_metrics.f90                         done.
nf_activation.f90                      done.
nf_datasets_mnist.f90                  done.
nf_datasets.f90                        done.
nf_io_binary.f90                       done.
nf_layer.f90                           failed.
nf_input3d_layer.f90                   failed.
nf_dense_layer.f90                     failed.
nf_input1d_layer.f90                   failed.
nf_datasets_submodule.f90              failed.
nf_datasets_mnist_submodule.f90        failed.
nf_flatten_layer.f90                   failed.
nf_conv2d_layer.f90                    failed.
[ 33%] Compiling...
f951.exe: Fatal Error: Reading module 'nf_activation.mod' at line 22 column 34: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'build\gfortran_2654F75F5833692A/nf_datasets_mnist.smod' at line 63 column 62: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'build\gfortran_2654F75F5833692A/nf_datasets.smod' at line 41 column 56: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'nf_activation.mod' at line 22 column 34: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'build\gfortran_2654F75F5833692A/nf_base_layer.mod' at line 17 column 21: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'build\gfortran_2654F75F5833692A/nf_base_layer.mod' at line 17 column 21: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'build\gfortran_2654F75F5833692A/nf_base_layer.mod' at line 17 column 21: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'build\gfortran_2654F75F5833692A/nf_base_layer.mod' at line 17 column 21: Unexpected EOF
compilation terminated.
<ERROR> Compilation failed for object " src_nf_nf_conv2d_layer.f90.o "
<ERROR> Compilation failed for object " src_nf_nf_datasets_mnist_submodule.f90.o "
<ERROR> Compilation failed for object " src_nf_nf_datasets_submodule.f90.o "
<ERROR> Compilation failed for object " src_nf_nf_dense_layer.f90.o "
<ERROR> Compilation failed for object " src_nf_nf_flatten_layer.f90.o "
<ERROR> Compilation failed for object " src_nf_nf_input1d_layer.f90.o "
<ERROR> Compilation failed for object " src_nf_nf_input3d_layer.f90.o "
<ERROR> Compilation failed for object " src_nf_nf_layer.f90.o "
<ERROR> stopping due to failed compilation
STOP 1
PS C:\Users\angel\neural-fortran> fpm build --profile release
nf_datasets_mnist_submodule.f90        failed.
nf_input1d_layer.f90                   failed.
nf_conv2d_layer.f90                    failed.
nf_datasets_submodule.f90              failed.
nf_dense_layer.f90                     failed.
nf_input3d_layer.f90                   failed.
nf_flatten_layer.f90                   failed.
nf_layer.f90                           failed.
[ 19%] Compiling...
f951.exe: Fatal Error: Reading module 'nf_activation.mod' at line 22 column 34: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'build\gfortran_2654F75F5833692A/nf_datasets_mnist.smod' at line 63 column 62: Unexpected EOF
compilation terminated.
f951.exe: Fatal Error: Reading module 'build\gfortran_2654F75F5833692A/nf_datasets.smod' at line 41 column 56: Unexpected EOF
compilation terminated.`

Could you help me fixing this issue?

**Comment by milancurcic** [comment]
Not clear to me yet why this happens. May be Windows-specific. What version of gfortran is this?

**Comment by ricor07** [comment]
14.2.0. I think it is the last one

**Comment by milancurcic** [comment]
OK, I just built and tested `main` with gfortran-14.2.0, albeit on Linux, with no issues.

Just to be sure, can you try removing the `build` directory and running `fpm build` again?

If the issue still appears the same way, let's try [building with CMake](https://github.com/modern-fortran/neural-fortran#building-with-cmake).

**Comment by ricor07** [comment]
Yes, already done, the issue persists.
Before trying FPM, i already tried Cmake. Now I don't have the logs but tomorrow (it's 11 pm here) i can post them here.

**Comment by milancurcic** [comment]
OK, cool, thanks. I think this points toward the mingw or its gfortran distribution, or how they're configured. I'm not familiar with compiling on Windows or mingw but I know others who are and we can ask.

When you get a chance, can you try compiling a simple hello world program, and then a module source file, with gfortran directly?

**Comment by ricor07** [comment]
Sure. I'll do it tomorrow morning (I think 5.00 am in Miami). Are you active also on Saturdays?

**Comment by milancurcic** [comment]
No rush and whenever it's convenient for you. I'm off until January 6 but read email and can respond on most days.

**Comment by ricor07** [comment]
Hello. The hello world program and the module source file run perfectly. I even tried building with cmake again but it stops reading some code chunks, as i posted before with fpm.  These are the logs i get with the make command:

[ 86%] Building Fortran object example/CMakeFiles/dense_mnist.dir/dense_mnist.f90.obj
C:\Users\angel\neural-fortran\example\dense_mnist.f90:3:74:

    3 |   use nf, only: dense, input, network, sgd, label_digits, load_mnist, corr
      |                                                                          1
Error: Unexpected USE statement at (1)
C:\Users\angel\neural-fortran\example\dense_mnist.f90:5:15:

    5 |   implicit none
      |               1
Error: Unexpected IMPLICIT NONE statement at (1)
C:\Users\angel\neural-fortran\example\dense_mnist.f90:7:16:

    7 |   type(network) :: net
      |                1
Error: Derived type 'network' at (1) is being used before it is defined
C:\Users\angel\neural-fortran\example\dense_mnist.f90:8:63:

    8 |   real, allocatable :: training_images(:,:), training_labels(:)
      |                                                               1
Error: Unexpected data declaration statement at (1)
C:\Users\angel\neural-fortran\example\dense_mnist.f90:9:67:

    9 |   real, allocatable :: validation_images(:,:), validation_labels(:)
      |                                                                   1
Error: Unexpected data declaration statement at (1)
C:\Users\angel\neural-fortran\example\dense_mnist.f90:10:26:

   10 |   integer :: n, num_epochs
      |                          1
Error: Unexpected data declaration statement at (1)
C:\Users\angel\neural-fortran\example\dense_mnist.f90:25:10:

   25 |   call net % print_info()
      |          1
Error: VARIABLE attribute of 'net' conflicts with PROCEDURE attribute at (1)
C:\Users\angel\neural-fortran\example\dense_mnist.f90:32:12:

   32 |     call net % train( &
      |            1
Error: VARIABLE attribute of 'net' conflicts with PROCEDURE attribute at (1)
C:\Users\angel\neural-fortran\example\dense_mnist.f90:44:29:

   44 |       output_metrics = net % evaluate(validation_images, label_digits(validation_labels), metric=corr())
      |                             1
Error: Symbol 'net' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:56:18:

   18 |   net = network([ &
      |               2
......
   56 |     type(network), intent(in out) :: net
      |                  1
Error: Type name 'network' at (1) conflicts with previously declared entity at (2), which has the same name
C:\Users\angel\neural-fortran\example\dense_mnist.f90:61:27:

   61 |       if (all(maxloc(net % predict(x(:,i))) == maxloc(y(:,i)))) then
      |                           1
Error: Symbol 'net' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:63:9:

   63 |       end if
      |         1
Error: Expecting END DO statement at (1)
C:\Users\angel\neural-fortran\example\dense_mnist.f90:55:28:

   55 |   real function accuracy(net, x, y)
      |                            1
Error: Symbol 'net' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:30:14:

   30 |   epochs: do n = 1, num_epochs
      |              1
Error: Symbol 'n' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:18:5:

   18 |   net = network([ &
      |     1
Error: Symbol 'net' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:23:12:

   23 |   num_epochs = 10
      |            1
Error: Symbol 'num_epochs' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:12:33:

   12 |   call load_mnist(training_images, training_labels, &
      |                                 1
Error: Symbol 'training_images' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:12:50:

   12 |   call load_mnist(training_images, training_labels, &
      |                                                  1
Error: Symbol 'training_labels' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:13:35:

   13 |                   validation_images, validation_labels)
      |                                   1
Error: Symbol 'validation_images' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:13:54:

   13 |                   validation_images, validation_labels)
      |                                                      1
Error: Symbol 'validation_labels' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:12:33:

   12 |   call load_mnist(training_images, training_labels, &
      |                                 1
Error: Symbol 'training_images' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:18:17:

   18 |   net = network([ &
      |                 1
Error: Function 'input' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:19:15:

   19 |     input(784), &
      |               1
Error: Function 'dense' at (1) has no IMPLICIT type
C:\Users\angel\neural-fortran\example\dense_mnist.f90:20:14:

   20 |     dense(30), &
      |              1
Error: Function 'dense' at (1) has no IMPLICIT type
mingw32-make[3]: *** [example\CMakeFiles\dense_mnist.dir\build.make:77: example/CMakeFiles/dense_mnist.dir/dense_mnist.f90.obj] Error 1
mingw32-make[2]: *** [CMakeFiles\Makefile2:735: example/CMakeFiles/dense_mnist.dir/all] Error 2
mingw32-make[1]: *** [Makefile:145: all] Error 2
mingw32-make[1]: Leaving directory 'C:/Users/angel/neural-fortran/build'

**Comment by ricor07** [comment]
Hello, I write just to confirm I am there. I already have WSL but i never used it. If you have time could you tell me what to do with it? I can review the issue on Saturday. Thanks and good evening

**Comment by milancurcic** [comment]
Thank you, no problem. I needed to move forward with dropout for a funded project.

The first step on WSL would be to find how to install git, gfortran, cmake, and fpm on it. If you tell me what OS it is (Ubuntu or something else?), I can provide more specifics.

**Comment by ricor07** [comment]
Hello, I'm happy to tell that now i can work on the repo. I see you added other commits that close #194 and the tests work. What do I have to work with?

**Comment by milancurcic** [comment]
Great to hear; for the time being, you can just review the code and see if everything makes sense so far. If yes, you can also try adding a small test (in test_dropout_layer.f90) for a backward pass.
